src/librustc_lint/builtin.rs

```diff
@@ -2087,6 +2087,82 @@ impl LintPass for InvalidNoMangleItems {
     }
 }
 
+#[derive(Clone, Copy)]
+pub struct MutableTransmutes;
+
+declare_lint! {
+    MUTABLE_TRANSMUTES,
+    Deny,
+    "transmuting from &T to &mut T is undefined behavior"
+}
+
+impl LintPass for MutableTransmutes {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(MUTABLE_TRANSMUTES)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &ast::Expr) {
+        use syntax::ast::DefId;
+        use syntax::abi::RustIntrinsic;
+        let msg = "transmuting from immutable to mutable is undefined behavior";
+        match get_transmute_from_to(cx, expr) {
+            Some((&ty::ty_rptr(_, from_mt), &ty::ty_rptr(_, to_mt))) => {
+                if to_mt.mutbl == ast::Mutability::MutMutable
+                    && from_mt.mutbl == ast::Mutability::MutImmutable {
+                    cx.span_lint(MUTABLE_TRANSMUTES, expr.span, msg);
+                }
+            }
+            _ => ()
+        }
+
+        fn get_transmute_from_to<'a, 'tcx>(cx: &Context<'a, 'tcx>, expr: &ast::Expr)
+            -> Option<(&'tcx ty::sty<'tcx>, &'tcx ty::sty<'tcx>)> {
+            if let ast::ExprPath(..) = expr.node {
+                if let DefFn(did, _) = ty::resolve_expr(cx.tcx, expr) {
+                    if def_id_is_transmute(cx, did) {
+                        let typ = ty::node_id_to_type(cx.tcx, expr.id);
```

Comment:
It seems it's needed to get to the `DefId`, so that we can determine if this method is actually `std::mem::transmute`, even if it's been imported as a different name. The `BareFnTy` is also needed to check the ABI.